src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java

```diff
@@ -0,0 +1,192 @@
+import io.vertx.core.Vertx;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.prebid.server.auction.model.Tuple2;
+import org.prebid.server.metric.MetricName;
+import org.prebid.server.metric.Metrics;
+import org.prebid.server.settings.CacheNotificationListener;
+import org.prebid.server.settings.model.StoredDataResult;
+import org.prebid.server.vertx.Initializable;
+import software.amazon.awssdk.core.BytesWrapper;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.ListObjectsRequest;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * <p>
+ * Service that periodically calls s3 for stored request updates.
+ * If refreshRate is negative, then the data will never be refreshed.
+ * <p>
+ * Fetches all files from the specified folders/prefixes in s3 and downloads all files.
+ */
+public class S3PeriodicRefreshService implements Initializable {
+
+    private static final String JSON_SUFFIX = ".json";
+
+    private static final Logger logger = LoggerFactory.getLogger(S3PeriodicRefreshService.class);
+
+    private final S3AsyncClient asyncClient;
+    private final String bucket;
+    private final String storedImpressionsDirectory;
+    private final String storedRequestsDirectory;
+    private final long refreshPeriod;
+    private final long timeout;
+    private final MetricName cacheType;
+    private final CacheNotificationListener cacheNotificationListener;
+    private final Vertx vertx;
+    private final Metrics metrics;
+    private final Clock clock;
+    private StoredDataResult lastResult;
+
+    public S3PeriodicRefreshService(S3AsyncClient asyncClient,
+                                    String bucket,
+                                    String storedRequestsDirectory,
+                                    String storedImpressionsDirectory,
+                                    long refreshPeriod,
+                                    long timeout,
+                                    MetricName cacheType,
+                                    CacheNotificationListener cacheNotificationListener,
+                                    Vertx vertx,
+                                    Metrics metrics,
+                                    Clock clock) {
+
+        this.asyncClient = asyncClient;
+        this.bucket = bucket;
+        this.storedRequestsDirectory = storedRequestsDirectory;
+        this.storedImpressionsDirectory = storedImpressionsDirectory;
+        this.refreshPeriod = refreshPeriod;
+        this.timeout = timeout;
```

Comment:
Please, add call to `Objects.requireNonNull` to all nullable fields initializations.